File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/6-Undirected Graph Representation/undirected_graph.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Adjacency list representation of undirected graph.

Note that parallel edges and self-loops are not allowed.
"""

__author__ = 'Ziang Lu'


class Vertex(object):
    def __init__(self, vtx_id):
        """
        Constructor with parameter.
        :param vtx_id: int
        """
        self._vtx_id = vtx_id
        self._my_edges = []

    @property
    def vtx_id(self):
        """
        Accessor of vtx_id.
        :return: int
        """
        return self._vtx_id

    def add_edge(self, new_edge):
        """
        Adds the given edge to this vertex.
        :param new_edge: Edge
        :return: None
        """
        # Check whether the input edge is None
        if new_edge is None:
            print('The input edge should not be None.')
            return
        # Check whether the input edge involves this vertex
        if new_edge.end1 != self and new_edge.end2 != self:
            print('The input edge should involve this vertex.')
            return

        self._my_edges.append(new_edge)

    def remove_edge(self, edge_to_remove):
        """
        Removes the given edge from this vertex.
        :param edge_to_remove: Edge
        :return: None
        """
        # Check whether the input edge is None
        if edge_to_remove is None:
            print('The input edge should not be None.')
            return
        # Check whether the input edge involves this vertex
        if edge_to_remove.end1 != self and edge_to_remove.end2 != self:
            print('The input edge should involve this vertex.')
            return

        self._my_edges.remove(edge_to_remove)

    def __repr__(self):
        """
        String representation of this vertex.
        :return:
        """
        return "Vertex #%d" % self._vtx_id

    def __eq__(self, other):
        """
        Equality test between this and the given vertex.
        :param other: Vertex
        :return: bool
        """
        return isinstance(other, Vertex) and self._vtx_id == other.vtx_id


class Edge(object):
    def __init__(self, end1, end2):
        """
        Constructor with parameter.
        :param end1: Vertex
        :param end2: Vertex
        """
        # Check whether the input endpoints are None
        if end1 is None or end2 is None:
            print('The endpoints should not be None.')
            return
        # Check whether the input endpoints are the same (self-loop)
        if end1.vtx_id == end2.vtx_id:
            print('The endpoints are the same (self-loop).')
            return

        self._end1 = end1
        self._end2 = end2

    @property
    def end1(self):
        """
        Accessor of end1.
        :return: Vertex
        """
        return self._end1

    @property
    def end2(self):
        """
        Accessor of end2.
        :return: Vertex
        """
        return self._end2

    def __repr__(self):
        """
        String representation of this edge.
        :return: str
        """
        return 'Edge between Vertex #%d and Vertex #%d' % \
            (self._end1.vtx_id, self._end2.vtx_id)


class AdjacencyList(object):
    def __init__(self):
        """
        Default constructor.
        """
        self._vtx_list = []
        self._edge_list = []

    def add_vtx(self, new_vtx_id):
        """
        Adds a new vertex to this graph.
        :param new_vtx_id: int
        :return: None
        """
        # Check whether the input vertex is repeated
        if self._find_vtx(vtx_id=new_vtx_id):
            print('The input vertex is repeated.')
            return

        new_vtx = Vertex(new_vtx_id)
        self._vtx_list.append(new_vtx)

    def _find_vtx(self, vtx_id):
        """
        Private helper function to find the given vertex in this adjacency list.
        :param vtx_id: int
        :return: Vertex
        """
        for vtx in self._vtx_list:
            if vtx.vtx_id == vtx_id:
                return vtx
        # Not found
        return None

    def add_edge(self, end1_id, end2_id):
        """
        Adds a new edge to this graph.
        :param end1_id: int
        :param end2_id: int
        :return: None
        """
        # Check whether the input vertices both exist
        end1, end2 = self._find_vtx(vtx_id=end1_id), \
            self._find_vtx(vtx_id=end2_id)
        if end1 is None or end2 is None:
            print("The input vertices don't both exist.")
            return
        # Check whether the edge to add already exists.
        if self._find_edge(end1_id=end1_id, end2_id=end2_id):
            print("The edge to add already exists.")
            return

        new_edge = Edge(end1, end2)
        end1.add_edge(new_edge=new_edge)
        end2.add_edge(new_edge=new_edge)
        self._edge_list.append(new_edge)

    def _find_edge(self, end1_id, end2_id):
        """
        Private helper function to find the edge connecting the given two
        vertices in this adjacency list.
        :param end1_id: int
        :param end2_id: int
        :return: Edge
        """
        for edge in self._edge_list:
            curr_end1_id, curr_end2_id = edge.end1.vtx_id, edge.end2.vtx_id
            if (curr_end1_id == end1_id and curr_end2_id == end2_id) or \
                    (curr_end1_id == end2_id and curr_end2_id == end1_id):
                return edge
        # Not found
        return None

    def remove_edge(self, end1_id, end2_id):
        """
        Removes an edge from this graph.
        :param end1_id: int
        :param end2_id: int
        :return: None
        """
        # Check whether the input vertices both exist
        end1, end2 = self._find_vtx(vtx_id=end1_id), \
            self._find_vtx(vtx_id=end2_id)
        if end1 is None or end2 is None:
            print("The input vertices don't both exist.")
            return
        # Check whether the edge to remove exists
        edge_to_remove = self._find_edge(end1_id=end1_id, end2_id=end2_id)
        if edge_to_remove is None:
            print("The edge to remove doesn't exist.")
            return

        end1.remove_edge(edge_to_remove)
        end2.remove_edge(edge_to_remove)
        self._edge_list.remove(edge_to_remove)

    def show_graph(self):
        """
        Shows this graph.
        :return:
        """
        print('The vertices are:')
        for vtx in self._vtx_list:
            print(vtx)
        print('The edges are:')
        for edge in self._edge_list:
            print(edge)

File: 3-Greedy Algorithms, Minimum Spanning Tree, and Dynamic Programming/1-Greedy Algorithm/2-Single-Vertex Shortest Paths/Java/undirected_graph/UndirectedGraph.java
package undirected_graph;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.PriorityQueue;

import graph.GraphInterface;

/**
 * Adjacency list representation of a undirected graph.
 *
 * Note that parallel edges are allowed, but not self-loops.
 * @author Ziang Lu
 */
public class UndirectedGraph implements GraphInterface {

    /**
     * Vertex list.
     */
    private final ArrayList<Vertex> vtxList;
    /**
     * Edge list.
     */
    private final ArrayList<UndirectedEdge> edgeList;

    /**
     * Default constructor.
     */
    public UndirectedGraph() {
        vtxList = new ArrayList<Vertex>();
        edgeList = new ArrayList<UndirectedEdge>();
    }

    @Override
    public void addVtx(int newVtxID) {
        // Check whether the input vertex is repeated
        if (findVtx(newVtxID) != null) {
            throw new IllegalArgumentException("The input vertex is repeated.");
        }

        Vertex newVtx = new Vertex(newVtxID);
        vtxList.add(newVtx);
    }

    /**
     * Private helper method to find the given vertex in this adjacency list.
     * @param vtxID vertex ID to look for
     * @return vertex if found, null if not found
     */
    private Vertex findVtx(int vtxID) {
        for (Vertex vtx : vtxList) {
            if (vtx.id() == vtxID) {
                return vtx;
            }
        }
        // Not found
        return null;
    }

    @Override
    public void removeVtx(int vtxID) {
        // Check whether the input vertex exists
        Vertex vtxToRemove = findVtx(vtxID);
        if (vtxToRemove == null) {
            throw new IllegalArgumentException("The input vertex doesn't exist.");
        }

        removeVtx(vtxToRemove);
    }

    /**
     * Private helper method to remove the given vertex from this graph.
     * @param vtxToRemove vertex to remove
     */
    private void removeVtx(Vertex vtxToRemove) {
        // Remove all the edges associated with the vertex to remove
        ArrayList<UndirectedEdge> edgesToRemove = vtxToRemove.edges();
        while (edgesToRemove.size() > 0) {
            UndirectedEdge edgeToRemove = edgesToRemove.get(0);
            removeEdge(edgeToRemove);
        }
        // Remove the vertex
        vtxList.remove(vtxToRemove);
    }

    @Override
    public void addEdge(int end1ID, int end2ID, double length) {
        // Check whether the input endpoints both exist
        Vertex end1 = findVtx(end1ID), end2 = findVtx(end2ID);
        if ((end1 == null) || (end2 == null)) {
            throw new IllegalArgumentException("The endpoints don't both exist.");
        }
        // Check whether the input endpoints are the same (self-loop)
        if (end1ID == end2ID) {
            throw new IllegalArgumentException("The endpoints are the same (self-loop).");
        }

        UndirectedEdge newEdge = new UndirectedEdge(end1, end2, length);
        addEdge(newEdge);
    }

    /**
     * Private helper method to add the given edge to this graph.
     * @param newEdge new edge
     */
    private void addEdge(UndirectedEdge newEdge) {
        Vertex end1 = newEdge.end1(), end2 = newEdge.end2();
        end1.addEdge(newEdge);
        end2.addEdge(newEdge);
        edgeList.add(newEdge);
    }

    @Override
    public void removeEdge(int end1ID, int end2ID) {
        // Check whether the input vertices both exist
        Vertex end1 = findVtx(end1ID), end2 = findVtx(end2ID);
        if ((end1 == null) || (end2 == null)) {
            throw new IllegalArgumentException("The input vertices don't both exist.");
        }
        // Check whether the edge to remove exists
        UndirectedEdge edgeToRemove = end1.getEdgeWithNeighbor(end2);
        if (edgeToRemove == null) {
            throw new IllegalArgumentException("The edge to remove doesn't exist.");
        }

        removeEdge(edgeToRemove);
    }

    /**
     * Private helper method to remove the given edge from this graph.
     * @param edgeToRemove edge to remove
     */
    private void removeEdge(UndirectedEdge edgeToRemove) {
        Vertex end1 = edgeToRemove.end1(), end2 = edgeToRemove.end2();
        end1.removeEdge(edgeToRemove);
        end2.removeEdge(edgeToRemove);
        edgeList.remove(edgeToRemove);
    }

    /**
     * Removes all the edges between a vertex pair from this graph.
     * @param end1ID endpoint1 ID
     * @param end2ID endpoint2 ID
     */
    public void removeEdgesBetweenPair(int end1ID, int end2ID) {
        try {
            while (true) {
                removeEdge(end1ID, end2ID);
            }
        } catch (IllegalArgumentException ex) {}
    }

    @Override
    public void showGraph() {
        System.out.println("The vertices are:");
        for (Vertex vtx : vtxList) {
            System.out.println(vtx);
        }
        System.out.println("The edges are:");
        for (UndirectedEdge edge : edgeList) {
            System.out.println(edge);
        }
    }

    @Override
    public HashMap<Integer, Double> dijkstraShortestPaths(int srcVtxID) {
        // Check whether the input source vertex exists
        Vertex srcVtx = findVtx(srcVtxID);
        if (srcVtx == null) {
            throw new IllegalArgumentException("The source vertex doesn't exist.");
        }

        // 1. Create a map between vertices and the shortest distances from the source vertex (X)
        HashMap<Integer, Double> shortestDistances = new HashMap<Integer, Double>();

        // Since we'll be using a lot of extract-min operations, we consider to use a heap to speed up the algorithm.

        // 2. Initialize the local minimum Dijkstra score of the source vertex to 0, and create a heap containing all
        //    the vertices not in X (V-X)
        srcVtx.setLocalMinScore(0);
        PriorityQueue<Vertex> vtxsToProcess = new PriorityQueue<Vertex>(vtxList);

        // 3. While X != V
        while (shortestDistances.size() < vtxList.size()) {
            // Among all directed crossing edges (v, w) with v in X (the map) and w in (V-X) (the heap), pick the edge
            // that minimizes A[v] + l_vw ("Dijkstra's Greedy Criterion")

            // Simply extract the vertex with the minimum Dijkstra score from the heap (V-X), which is exactly w*
            // In other words, in the tournament's second round, each vertex proposes its "local" winner score, and the
            // heap (V-X) chooses the "overall" winner score.
            Vertex wStar = vtxsToProcess.poll();
            double minScore = wStar.localMinScore();

            // Put w* and the corresponding shortest distance to the map (X)
            shortestDistances.put(wStar.id(), minScore);

            // Extracting one vertex from the heap (V-X) may influence some local minimum Dijkstra scores of vertices
            // that are still in the heap (V-X).
            // Vertices that are not connected to w* and are still in the heap (V-X) won't be influenced.
            // => The local minimum Dijkstra scores of vertices that are connected to w* and are still in the heap (V-X)
            //    may drop down.

            // Update the local minimum Dijkstra scores for the vertices if necessary
            for (UndirectedEdge edgeFromWStar : wStar.edges()) {
                // Find the neighbor
                Vertex neighbor = null;
                if (edgeFromWStar.end1() == wStar) { // endpoint2 is the neighbor.
                    neighbor = edgeFromWStar.end2();
                } else { // endpoint1 is the neighbor
                    neighbor = edgeFromWStar.end1();
                }
                // Check whether the neighbor of w* is still in the heap (V-X)
                // This is equivalent to checking whether this neighbor is not in the map (X).
                // The reason we choose to check the latter condition is because searching is faster in hash table
                // (O(1)).
                if (!shortestDistances.containsKey(neighbor.id())) {
                    // Check whether the local minimum Dijkstra score of this neighbor needs to be updated
                    double newScore = minScore + edgeFromWStar.length();
                    if (newScore < neighbor.localMinScore()) {
                        // Remove this neighbor from the heap (V-X)
                        vtxsToProcess.remove(neighbor);
                        // Update its local minimum Dijkstra score
                        neighbor.setLocalMinScore(newScore);
                        // Put this neighbor back to the heap (V-X)
                        vtxsToProcess.offer(neighbor);
                    }
                }
            }
        }
        return shortestDistances;
    }

}
